sdk/core/azure-core-http-netty/src/main/java/com/azure/core/http/netty/ReactorNettyHttpResponseBase.java

```diff
@@ -31,8 +42,136 @@ public final String getHeaderValue(String name) {
 
     @Override
     public final HttpHeaders getHeaders() {
-        HttpHeaders headers = new HttpHeaders();
-        reactorNettyResponse.responseHeaders().forEach(e -> headers.put(e.getKey(), e.getValue()));
+        if (headers == null) {
+            headers = new NettyHttpHeaders(reactorNettyResponse.responseHeaders());
+        }
         return headers;
     }
+
+    static class NettyHttpHeaders extends HttpHeaders {
+        private final io.netty.handler.codec.http.HttpHeaders nettyHeaders;
+        private Map<String, String> map;
+
+        NettyHttpHeaders(io.netty.handler.codec.http.HttpHeaders nettyHeaders) {
+            this.nettyHeaders = nettyHeaders;
+        }
+
+        private String formatKey(final String key) {
+            return key == null ? null : key.toLowerCase(Locale.ROOT);
+        }
+
+        @Override
+        public int getSize() {
+            return nettyHeaders.size();
+        }
+
+        /* This replaces any current value with the given name */
+        @Override
+        public HttpHeaders put(String name, String value) {
+            if (name == null) {
+                return this;
+            }
+
+            name = formatKey(name);
+            if (value == null) {
+                remove(name);
+            } else {
+                nettyHeaders.set(name, value);
+            }
+            return this;
+        }
+
+        HttpHeaders add(String name, String value) {
+            if (name == null) {
+                return this;
+            }
+
+            name = formatKey(name);
+            if (value == null) {
+                remove(name);
+            } else {
+                nettyHeaders.add(formatKey(name), value);
+            }
+            return this;
+        }
+
+        @Override
+        public HttpHeader get(String name) {
+            name = formatKey(name);
+            if (nettyHeaders.contains(name)) {
+                return new NettyHttpHeader(this, name, nettyHeaders.get(name));
+            }
+            return null;
+        }
+
+        @Override
+        public HttpHeader remove(String name) {
+            name = formatKey(name);
+            HttpHeader header = get(name);
+            nettyHeaders.remove(name);
+            return header;
+        }
+
+        @Override
+        public String getValue(String name) {
+            return nettyHeaders.get(formatKey(name));
+        }
+
+        @Override
+        public String[] getValues(String name) {
+            return nettyHeaders.getAll(formatKey(name)).toArray(new String[] { });
+        }
+
+        @Override
+        public Map<String, String> toMap() {
```

Comment:
This changes the existing behavior of `toMap` in `HttpHeaders`. Previously, `toMap()` returned a new copy of headers and any changes to the copy would not affect another `toMap` caller. 

Also, the previous implementation allowed `put` operations on the returned map as it used `HashMap`. This change will throw `UnsupportedOperationException` because it uses `AbstractMap`.  We should consider making this an unmodifiable map if we don't want to support changes to the map.

This change seems like the right behavior but just noting that this could potentially be a behavioral breaking change.

Reply:
Agree, we should document that toMap returns an immutable copy, and make this change in our other implementation.

Reply:
Yeah, agree, since this is a map of response headers returned by the service, there's no need for this map to be mutable.